docs/source/users/configuration.rst
Configuration
*************

Mightstone behavior is defined by configuration. You can configure your mightstone by three means:

* Programmatically with a ``MightstoneSettings`` instance passed to your ``Mightstone`` application instance
* Through environment variables, for instance ``MIGHTSTONE_APPNAME`` to alter ``appname`` key
* Through a configuration file, with 3 means of defining

  * In CLI by passing a ``--config`` parameter to the command line client
  * Via ``MIGHTSTONE_CONFIG_FILE`` environment variable
  * Passively, by storing your configuration file in an standard location.

Configuration
=============

=========================== ======================================  ====== ======================== ===========
Path                        Environment Variable                    Type   default                  description
=========================== ======================================  ====== ======================== ===========
``appname``                 ``MIGHTSTONE_APPNAME``                  string ``mightstone``           Your application name, it will be used to build your storage and cache paths on the filesystem.
=========================== ======================================  ====== ======================== ===========



Storage
-------

=========================== ======================================  ====== ====================================== ===========
Path                        Environment Variable                    Type   default                                description
=========================== ======================================  ====== ====================================== ===========
``storage.implementation``  ``MIGHTSTONE_STORAGE_IMPLEMENTATION``   string ``local``                              Either ``local`` for a locally hosted mongod, ``motor`` for a remote mongodb connection.
``storage.uri``             ``MIGHTSTONE_STORAGE_URI``              string                                        Mongodb dsn (``mongodb://login:password@host/db``...), only used if ``motor`` implementation
``storage.directory``       ``MIGHTSTONE_STORAGE_DIRECTORY``        string ``<user_data>/Mightstone/data/mongo``  Data storage directory, only used if ``local`` implementation
``storage.database``        ``MIGHTSTONE_STORAGE_DATABASE``         string ``mightstone``                         The mongo database
=========================== ======================================  ====== ====================================== ===========

By default, Mighstone will use a jerry-rigged storage mechanism based on ``pymongo_inmemory``. This will not work with multiple concurrent instances of Mightstone tough.
There was an aborted attempt to use ``beanita`` / ``mongita`` backend, but we are still in dire need of features not implemented yet, and some bugfixes.

Http
----

=========================== ======================================  ============= ========================= ===========
Path                        Environment Variable                    Type          default                   description
=========================== ======================================  ============= ========================= ===========
``http.cache.persist``      ``MIGHTSTONE_CACHE_PERSIST``            bool          ``True``                  Use memory (transient) or filesystem (persistent) cache
``http.cache.directory``    ``MIGHTSTONE_CACHE_DIRECTORY``          string        ``<user_cache>/http``     The directory to store HTTP cache, only used of ``persist`` is ``true``
``http.cache.methods``      ``MIGHTSTONE_CACHE_METHODS``            list[string]  ``[GET]``                 HTTP verbs to cache
``http.cache.status``       ``MIGHTSTONE_CACHE_STATUS``             list[int]     ``[200,203,300,301,308]`` HTTP status code to cache
=========================== ======================================  ============= ========================= ===========


Where to store your configuration ?
===================================

Mightstone will search configuration file in this order:

* ``<current_directory>/mightstone.yaml``
* ``<current_directory>/mightstone.yml``
* ``<current_directory>/mightstone.json``
* ``<current_directory>/mightstone.toml``
* ``<user_config_directory>/mightstone.yaml``
* ``<user_config_directory>/mightstone.yml``
* ``<user_config_directory>/mightstone.json``
* ``<user_config_directory>/mightstone.toml``

Mighstone relies on `AppDirs <https://github.com/ActiveState/appdirs>`_ package to resolve user config directory:

::

    Mac OS X:               ~/Library/Application Support/Mightstone
    Unix:                   ~/.config/Mightstone     # or in $XDG_CONFIG_HOME, if defined
    Win XP (not roaming):   C:\Documents and Settings\<username>\Application Data\Mightstone\Mightstone
    Win XP (roaming):       C:\Documents and Settings\<username>\Local Settings\Application Data\Mightstone\Mightstone
    Win 7  (not roaming):   C:\Users\<username>\AppData\Local\Mightstone\Mightstone
    Win 7  (roaming):       C:\Users\<username>\AppData\Roaming\Mightstone\Mightstone


CHANGELOG.md
# CHANGELOG



## v0.8.5 (2024-06-03)

### Fix

* fix: dependency lock was incorrect ([`45e8da5`](https://github.com/Guibod/mightstone/commit/45e8da5e616834cdf7a2ac38fbc2f25938df1a80))

* fix: python 3.12 support, in dependencies ([`7502564`](https://github.com/Guibod/mightstone/commit/7502564f10c0c50b62c5fea90c7ad3d6c0fe81b9))


## v0.8.4 (2024-06-03)

### Fix

* fix: pypi publish step should now release the project ([`4fd614e`](https://github.com/Guibod/mightstone/commit/4fd614e71abc6ed9dd2a091e1c8f06a1002de2bc))


## v0.8.3 (2024-06-03)

### Fix

* fix: using pypa trusted release mechanism to publish on pypi (properly) ([`cd28493`](https://github.com/Guibod/mightstone/commit/cd28493d54b0cd39aacbc8edb556585432a09e18))

* fix: using pypa trusted release mechanism to publish on pypi ([`09a08ac`](https://github.com/Guibod/mightstone/commit/09a08ac2132835577c105f0978322671e167594d))


## v0.8.2 (2024-06-03)

### Build

* build: attempt to restore build sequence on release ([`4321db8`](https://github.com/Guibod/mightstone/commit/4321db8f8508b2856c344f8a170904a08df94fad))

### Chore

* chore: moved from flake8 to ruff ([`77238d9`](https://github.com/Guibod/mightstone/commit/77238d9b14f49c58e5082a6da4507056847e8556))

* chore: added tox support as part of pyproject configuration ([`1030b2b`](https://github.com/Guibod/mightstone/commit/1030b2b1c90e9750ab5499b87dcbda88a3cfe096))

* chore: removed poetry.yml file ([`1523714`](https://github.com/Guibod/mightstone/commit/1523714c7f78d836af619401dee85bf9d654a0f4))

### Fix

* fix: release will happen on push on main branch, not master ([`d3aa585`](https://github.com/Guibod/mightstone/commit/d3aa585284a88f1b8c5890c31701e012ea809940))

* fix: restored support to python 3.12

by removing dependency to python-dependency-injection for injector ([`79ddd90`](https://github.com/Guibod/mightstone/commit/79ddd90290603261fcfca7deea0574c287241bb2))

* fix: fixed python dependency properly ([`c952fa3`](https://github.com/Guibod/mightstone/commit/c952fa323797a505ca65e61084b7143621a9a475))

* fix: build process is now splitted into: ci, semantic-release and on-release ([`d74a45e`](https://github.com/Guibod/mightstone/commit/d74a45ea51784700ee468ed5e2035a954dfed2f2))

### Unknown

* tests: CI will run remote tests ([`731905d`](https://github.com/Guibod/mightstone/commit/731905d1ca8a9d2979f806b8e09cccf75db3219e))


## v0.8.1 (2024-05-31)

### Chore

* chore: updated mypy and updated types ([`5524dd7`](https://github.com/Guibod/mightstone/commit/5524dd781ddbc220440b4f3bfa37e693270fe017))

* chore: silenced annoying PytestUnraisableExceptionWarning ([`b7e0bc1`](https://github.com/Guibod/mightstone/commit/b7e0bc1c0a6a21e7d364985a63bb9cc4a132de27))

* chore: applied new black rules to the code ([`8690b52`](https://github.com/Guibod/mightstone/commit/8690b52db051ecae317e693bdd201bc1687524c3))

* chore(deps-dev): bump black from 23.12.1 to 24.3.0

Bumps [black](https://github.com/psf/black) from 23.12.1 to 24.3.0.
- [Release notes](https://github.com/psf/black/releases)
- [Changelog](https://github.com/psf/black/blob/main/CHANGES.md)
- [Commits](https://github.com/psf/black/compare/23.12.1...24.3.0)

---
updated-dependencies:
- dependency-name: black
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] &lt;support@github.com&gt; ([`89fc0c1`](https://github.com/Guibod/mightstone/commit/89fc0c14e684016013c1d5734211fda02c2dec04))

* chore(deps-dev): bump pyinstaller from 4.5.1 to 5.13.1

Bumps [pyinstaller](https://github.com/pyinstaller/pyinstaller) from 4.5.1 to 5.13.1.
- [Release notes](https://github.com/pyinstaller/pyinstaller/releases)
- [Changelog](https://github.com/pyinstaller/pyinstaller/blob/develop/doc/CHANGES.rst)
- [Commits](https://github.com/pyinstaller/pyinstaller/compare/v4.5.1...v5.13.1)

---
updated-dependencies:
- dependency-name: pyinstaller
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] &lt;support@github.com&gt; ([`3cb34ee`](https://github.com/Guibod/mightstone/commit/3cb34ee6b3cdf71180c6ea8f974353af99714dd7))

* chore: build process lints before tests ([`a0c53d0`](https://github.com/Guibod/mightstone/commit/a0c53d0332de75ef24b2b0b5da9c6471be1be9b8))

* chore(deps): bump pillow from 9.5.0 to 10.3.0

Bumps [pillow](https://github.com/python-pillow/Pillow) from 9.5.0 to 10.3.0.
- [Release notes](https://github.com/python-pillow/Pillow/releases)
- [Changelog](https://github.com/python-pillow/Pillow/blob/main/CHANGES.rst)
- [Commits](https://github.com/python-pillow/Pillow/compare/9.5.0...10.3.0)

---
updated-dependencies:
- dependency-name: pillow
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] &lt;support@github.com&gt; ([`b9a94f7`](https://github.com/Guibod/mightstone/commit/b9a94f7b5e0e18d9ae1358f1d5d178f3a8451075))

* chore: readme don’t link to another project for issues, centered the logo ([`5ffd878`](https://github.com/Guibod/mightstone/commit/5ffd878f82cff56a3484c732b3d27567cf7f6c58))

### Documentation

* docs: fixed link to issue report ([`eba9be6`](https://github.com/Guibod/mightstone/commit/eba9be627cd6101327f4b112d8052603a22c2838))

### Fix

* fix: custom types are properly deserialed through pydantic 2 ([`0ab6c22`](https://github.com/Guibod/mightstone/commit/0ab6c221d0e8594fd09494fce91fda4d9b69ff60))


## v0.8.0 (2024-05-31)

### Chore

* chore: linting issues ([`d4cc2f9`](https://github.com/Guibod/mightstone/commit/d4cc2f92f1f6b044a17553af2c7412258285c495))

* chore: fixed minor typing issues after mypy upgrade ([`7bf3292`](https://github.com/Guibod/mightstone/commit/7bf329227f4a023a9cffd615d10d9a5ea63e588d))

* chore: upgraded dependency to aiofiles ([`d512acb`](https://github.com/Guibod/mightstone/commit/d512acb2406d8db50de67b6ff16d4c15c648496e))

* chore: remove dependency to aiosqlite ([`3c2de6b`](https://github.com/Guibod/mightstone/commit/3c2de6b54341559b3bfa133ab9b59b5ef6a67609))


src/mightstone/config.py
import json
import logging
import os
import pathlib
from enum import Enum
from typing import Literal, Optional, Tuple, Type, Union
from slugify import slugify

import toml
import yaml
from pydantic.networks import MongoDsn
from pydantic_settings import (
    BaseSettings,
    InitSettingsSource,
    JsonConfigSettingsSource,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
    TomlConfigSettingsSource,
    YamlConfigSettingsSource,
)

from mightstone.core import MightstoneError

logger = logging.getLogger("mightstone")


class DbImplem(str, Enum):
    LOCAL = "local"
    MOTOR = "motor"
    FAKE = "fake"


class IjsonEnum(str, Enum):
    YAJL2_C = "yajl2_c"
    YAJL2_CFFI = "yajl2_cffi"
    YAJL2 = "yajl2"
    YAJL = "yajl"
    PYTHON = "python"


class HttpCacheSettings(BaseSettings):
    persist: bool = True
    directory: Optional[pathlib.Path] = None
    methods: list[str] = ["GET"]
    status: list[int] = [200, 203, 300, 301, 308]


class HttpSettings(BaseSettings):
    cache: HttpCacheSettings = HttpCacheSettings()


class InMemorySettings(BaseSettings):
    implementation: Literal[DbImplem.LOCAL] = DbImplem.LOCAL
    directory: Optional[pathlib.Path] = None
    database: str = "mightstone"


class MotorSettings(BaseSettings):
    implementation: Literal[DbImplem.MOTOR] = DbImplem.MOTOR
    uri: MongoDsn
    database: str = "mightstone"


class FakeSettings(BaseSettings):
    implementation: Literal[DbImplem.FAKE] = DbImplem.FAKE
    database: str = "mightstone"


class MightstoneSettings(BaseSettings):
    appname: str = "Mightstone"
    storage: Union[InMemorySettings, MotorSettings, FakeSettings] = InMemorySettings()
    http: HttpSettings = HttpSettings()
    ijson: IjsonEnum = IjsonEnum.PYTHON

    model_config = SettingsConfigDict(
        {
            "env_prefix": "mightstone_",
            "env_nested_delimiter": "__",
            "env_file_encoding": "utf-8",
        }
    )

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: Type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> Tuple[PydanticBaseSettingsSource, ...]:
        return (
            init_settings,
            SettingsSourceGenerator("mightstone").build(settings_cls),
            env_settings,
            dotenv_settings,
            file_secret_settings,
        )


class SettingsSourceGenerator:
    def __init__(
        self, app_name: str = "mightstone", env_variable: Optional[str] = None
    ):
        self.app_name = slugify(app_name, separator="_")
        if not env_variable:
            env_variable = f"{self.app_name.upper()}_CONFIG_FILE"
        self.env_variable = env_variable

    def candidate_paths(
        self,
    ) -> list[pathlib.Path]:
        candidate = os.getenv(self.env_variable)
        if candidate:
            return [pathlib.Path(candidate)]

        return [
            pathlib.Path(path).joinpath(f"{self.app_name}.{extension}")
            for path in [os.getcwd(), os.path.expanduser("~")]
            for extension in ["yaml", "yml", "json", "toml"]
        ]

    @classmethod
    def load_file(cls, filepath, encoding):
        try:
            with open(filepath, "r", encoding=encoding) as f:
                _, extension = os.path.splitext(filepath)
                if extension.lower() in [".yml", ".yaml"]:
                    return yaml.safe_load(f)

                if extension.lower() == ".toml":
                    return toml.load(f)

                return json.load(f)
        except Exception as e:
            raise MightstoneError("Unable to parse %s" % filepath) from e

    def build(self, settings_cls: Type[BaseSettings]) -> Union[
        InitSettingsSource,
        JsonConfigSettingsSource,
        YamlConfigSettingsSource,
        TomlConfigSettingsSource,
    ]:
        for candidate in self.candidate_paths():
            if not candidate.exists():
                continue

            _, extension = os.path.splitext(candidate)
            try:
                if extension.lower() in [".yml", ".yaml"]:
                    return YamlConfigSettingsSource(settings_cls, candidate)

                if extension.lower() == ".toml":
                    return TomlConfigSettingsSource(settings_cls, candidate)

                return JsonConfigSettingsSource(settings_cls, candidate)
            except MightstoneError as e:
                logger.fatal("Unable to parse configuration from %s, %s", candidate, e)

        return InitSettingsSource(settings_cls, {})
